perfmetrics/scripts/job_files/seq_rand_read_write.fio

```diff
@@ -79,6 +82,7 @@ directory=gcs/50mb
 filesize=50M
 rw=write
 numjobs=40
+fsync=50
```

Comment:
can we use fsync_on_close https://fio.readthedocs.io/en/latest/fio_doc.html ? (not sure if fio closes file after writing complete file. if it closes then we can use, otherwise fsync=filesize/bs is fine)

Reply:
In my tests I found that fio always closes each file at the end of the test job irrespective of the value of `end_fsync` or `fsync_on_close` or of `fsync` . All 3 of these variable contain how frequently and where all syncFile operation is created.

Below is a log from a test job with 1filesize=2MB, bs=1MB, fsync=2` config.

```
time="15/12/2023 11:42:11.739023" severity=TRACE msg="fuse_debug: Op 0x00000204        connection.go:415] <- WriteFile (inode 4, PID 179987, handle 19, offset 0, 1048576 bytes)"
time="15/12/2023 11:42:13.276024" severity=TRACE msg="fuse_debug: Op 0x00000206        connection.go:415] <- WriteFile (inode 4, PID 179987, handle 19, offset 1048576, 1048576 bytes)"
time="15/12/2023 11:42:13.277153" severity=TRACE msg="fuse_debug: Op 0x0000020a        connection.go:415] <- SyncFile (inode 4, PID 179987)"
time="15/12/2023 11:42:13.509829" severity=TRACE msg="gcs: Req             0x33: <- CreateObject(\"fio_test/write.0.0\")"
time="15/12/2023 11:42:16.231577" severity=TRACE msg="gcs: Req             0x33: -> CreateObject(\"fio_test/write.0.0\") (2.721737186s): OK"
time="15/12/2023 11:42:16.232701" severity=TRACE msg="fuse_debug: Op 0x0000020c        connection.go:415] <- WriteFile (inode 4, PID 179987, handle 19, offset 0, 1048576 bytes)"
time="15/12/2023 11:42:17.811405" severity=TRACE msg="fuse_debug: Op 0x0000020e        connection.go:415] <- WriteFile (inode 4, PID 179987, handle 19, offset 1048576, 1048576 bytes)"
time="15/12/2023 11:42:17.812404" severity=TRACE msg="fuse_debug: Op 0x00000212        connection.go:415] <- SyncFile (inode 4, PID 179987)"
time="15/12/2023 11:42:18.029053" severity=TRACE msg="gcs: Req             0x36: <- CreateObject(\"fio_test/write.0.0\")"
time="15/12/2023 11:42:20.764991" severity=TRACE msg="gcs: Req             0x36: -> CreateObject(\"fio_test/write.0.0\") (2.735924437s): OK"
time="15/12/2023 11:42:20.766265" severity=TRACE msg="fuse_debug: Op 0x00000214        connection.go:415] <- WriteFile (inode 4, PID 179987, handle 19, offset 0, 1048576 bytes)"
time="15/12/2023 11:42:22.348341" severity=TRACE msg="fuse_debug: Op 0x00000216        connection.go:415] <- WriteFile (inode 4, PID 179987, handle 19, offset 1048576, 1048576 bytes)"
time="15/12/2023 11:42:22.349319" severity=TRACE msg="fuse_debug: Op 0x0000021a        connection.go:415] <- SyncFile (inode 4, PID 179987)"
time="15/12/2023 11:42:22.566568" severity=TRACE msg="gcs: Req             0x39: <- CreateObject(\"fio_test/write.0.0\")"
time="15/12/2023 11:42:25.290687" severity=TRACE msg="gcs: Req             0x39: -> CreateObject(\"fio_test/write.0.0\") (2.724108325s): OK"
time="15/12/2023 11:42:25.291813" severity=TRACE msg="fuse_debug: Op 0x0000021c        connection.go:415] <- WriteFile (inode 4, PID 179987, handle 19, offset 0, 1048576 bytes)"
time="15/12/2023 11:42:26.947532" severity=TRACE msg="fuse_debug: Op 0x0000021e        connection.go:415] <- WriteFile (inode 4, PID 179987, handle 19, offset 1048576, 1048576 bytes)"
time="15/12/2023 11:42:26.948286" severity=TRACE msg="fuse_debug: Op 0x00000222        connection.go:415] <- SyncFile (inode 4, PID 179987)"
time="15/12/2023 11:42:27.166873" severity=TRACE msg="gcs: Req             0x3c: <- CreateObject(\"fio_test/write.0.0\")"
time="15/12/2023 11:42:29.888007" severity=TRACE msg="gcs: Req             0x3c: -> CreateObject(\"fio_test/write.0.0\") (2.721121956s): OK"
time="15/12/2023 11:42:29.889243" severity=TRACE msg="fuse_debug: Op 0x00000224        connection.go:415] <- WriteFile (inode 4, PID 179987, handle 19, offset 0, 1048576 bytes)"
time="15/12/2023 11:42:31.394842" severity=TRACE msg="fuse_debug: Op 0x00000226        connection.go:415] <- WriteFile (inode 4, PID 179987, handle 19, offset 1048576, 1048576 bytes)"
time="15/12/2023 11:42:31.395657" severity=TRACE msg="fuse_debug: Op 0x0000022a        connection.go:415] <- SyncFile (inode 4, PID 179987)"
time="15/12/2023 11:42:31.609895" severity=TRACE msg="gcs: Req             0x3f: <- CreateObject(\"fio_test/write.0.0\")"
time="15/12/2023 11:42:34.338954" severity=TRACE msg="gcs: Req             0x3f: -> CreateObject(\"fio_test/write.0.0\") (2.729049447s): OK"
time="15/12/2023 11:42:34.339822" severity=TRACE msg="fuse_debug: Op 0x0000022c        connection.go:415] <- FlushFile (inode 4, PID 179987)"
time="15/12/2023 11:42:34.340059" severity=TRACE msg="fuse_debug: Op 0x0000022e        connection.go:415] <- ReleaseFileHandle (PID 0)"
```

Reply:
So, fsync_on_close isn't required in this case.

Reply:
It's not about whether fsync_on_close is required or not. It's about whether to simply use fsync_on_close and not have fsync specified for each job or use fsync in every job (as you have done currently).

IIUC from your logs, fio opens the files at the start of job and only closes at the end of job (i.e. after configured runtime). Because of this reason, we can't use fsync_on_close.

Reply:
Ok, I guess I misunderstood the question. The question is really about whether `fsync_on_close` alone could do the job of having to calculate and specify `fsync=filesize/bs` for each job separately. Yeah, I myself had the same thought earlier.

> Because of this reason, we can't use fsync_on_close.

Yeah. `Close` happens only once in fio even if a file is written multiple times. So, it will create a fake use case for us and will overestimate write-bandwidth by skipping necessary sync.